python/lsst/pipe/base/execution_reports.py

```diff
@@ -0,0 +1,438 @@
+                dataset_type_report = DatasetTypeExecutionReport()
+                self.output_datasets[output_ref.datasetType.name] = dataset_type_report
+            if output_ref.id not in refs[output_ref.datasetType.name]:
+                dataset_type_report.handle_missing_dataset(output_ref, failed, status_graph)
+            else:
+                dataset_type_report.handle_produced_dataset(output_ref, status_graph)
+
+    def to_summary_dict(self, butler: Butler, logs: bool = True) -> dict[str, Any]:
+        """Summarize the results of the TaskExecutionReport in a dictionary.
+
+        Parameters
+        ----------
+        butler: `lsst.daf.butler.Butler`
+            The Butler used for this report.
+        logs: `bool`
+            Store the logs in the summary dictionary.
+
+        Returns
+        -------
+        A dictionary containing:
+            outputs: `dict`
+                A dictionary summarizing the DatasetTypeExecutionReport for
+                each DatasetType associated with the task
+            failed_quanta: `dict`
+                A dictionary of quanta which failed and their dataIDs by
+                quantum graph node id
+            n_quanta_blocked: `int`
+                The number of quanta which failed due to upstream failures.
+        """
+        failed_quanta = {}
+        for node_id, log_ref in self.failed.items():
+            quantum_info: dict[str, Any] = {"data_id": log_ref.dataId.byName()}
+            if logs:
+                try:
+                    log = butler.get(log_ref)
+                except LookupError:
+                    quantum_info["error"] = []
+                else:
+                    quantum_info["error"] = [
+                        record.message for record in log if record.levelno >= logging.ERROR
+                    ]
+            failed_quanta[str(node_id)] = quantum_info
+        return {
+            "outputs": {name: r.to_summary_dict() for name, r in self.output_datasets.items()},
+            "failed_quanta": failed_quanta,
+            "n_quanta_blocked": len(self.failed_upstream),
+        }
+
+    def __str__(self) -> str:
+        """Return a count of the failed and failed_upstream tasks in the
+        TaskExecutionReport.
+        """
+        return f"failed: {len(self.failed)}\nfailed upstream: {len(self.failed_upstream)}\n"
+
+
+@dataclasses.dataclass
+class QuantumGraphExecutionReport:
+    """A report on the execution of a quantum graph.
+
+    Report the detailed status of each failure; whether tasks were not run,
+    data is missing from upstream failures, or specific errors occurred during
+    task execution (and report the errors). Contains a count of expected,
+    produced DatasetTypes for each task. This report can be output as a
+    dictionary or a yaml file.
+
+    Parameters
+    ----------
+    tasks: `dict`
+        A dictionary of TaskExecutionReports by pipetask
```

Comment:
```suggestion
        A dictionary of TaskExecutionReports by task label.
```

Formally `pipetask` is the command-line tool only, but we don't want to say `PipelineTask` here either if we want to be precise, since you could configure the same task multiple times in one pipeline, and it's really one labeled configuration we're talking about here.